File: xdm-core/src/main/java/xdm/core/media/parser/dash/MatcherResult.java
package xdm.core.media.parser.dash;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;

public class MatcherResult {
  private List<String> groups = new ArrayList<>();
  private Map<String, String> namedGroups = new HashMap<>();

  public String group(int index) {
    return groups.get(index);
  }

  public String group(String name) {
    return namedGroups.get(name);
  }

  private MatcherResult() {}

  public static MatcherResult toMatcherResult(Matcher matcher) {
    List<String> groups = new ArrayList<>();
    Map<String, String> namedGroups = new HashMap<>();
    for (int i = 0; i < matcher.groupCount(); i++) {
      groups.add(matcher.group(i));
    }
    namedGroups.put("timedx", matcher.group("timedx"));
    namedGroups.put("timedigits", matcher.group("timedigits"));
    namedGroups.put("numdx", matcher.group("numdx"));
    namedGroups.put("numdigits", matcher.group("numdigits"));
    MatcherResult matcherResult = new MatcherResult();
    matcherResult.groups = groups;
    matcherResult.namedGroups = namedGroups;
    return matcherResult;
  }
}

File: xdm-core/src/main/java/xdm/core/downloaders/web/streaming/manifest/dash/MatcherResult.kt
package xdm.core.downloaders.web.streaming.manifest.dash

import java.util.regex.Matcher

class MatcherResult private constructor() {
    private var groups: List<String> = ArrayList()
    private var namedGroups: Map<String, String> = HashMap()

    fun group(index: Int): String {
        return groups[index]
    }

    fun group(name: String): String? {
        return namedGroups[name]
    }

    companion object {
        fun toMatcherResult(matcher: Matcher): MatcherResult {
            val groups: MutableList<String> = ArrayList()
            val namedGroups: MutableMap<String, String> = HashMap()
            for (i in 0..<matcher.groupCount()) {
                groups.add(matcher.group(i))
            }
            namedGroups["timedx"] = matcher.group("timedx")
            namedGroups["timedigits"] = matcher.group("timedigits")
            namedGroups["numdx"] = matcher.group("numdx")
            namedGroups["numdigits"] = matcher.group("numdigits")
            val matcherResult = MatcherResult()
            matcherResult.groups = groups
            matcherResult.namedGroups = namedGroups
            return matcherResult
        }
    }
}
